### ISSUE TITLE
AR miss circular dependency / wrong has_many behaviour with inheritance

### ISSUE THREAD

**Issue by shlajin** [rails/rails#22548]
Hello!
ActiveRecord does not recognise circular dependencies if sub-classes laying inside the folders. If that's correct (not a circular dependency), then ActiveRecord breaks `has_many` defined in superclass. Let me show. (I've tested this code with Rails 4.2.5.)

``` ruby
class Animal < ActiveRecord::Base
  SUB_TYPES = [Cat].freeze
  has_many :abilities
end
```

``` ruby
class Ability < ActiveRecord::Base
  belongs_to :animal
end
```

``` ruby
class Cat < Animal
  has_many :toys, foreign_key: :animal_id
end
```

``` ruby
class Toy < ActiveRecord::Base
  belongs_to :animal
end
```

When I try to run `Cat.first.abilities` this code throws

```
RuntimeError: Circular dependency detected while autoloading constant Cat
```

And it's perfectly fine. But if I move `Cat` into `animal` folder

``` ruby
class Animal < ActiveRecord::Base
  SUB_TYPES = [Animal::Cat].freeze
  has_many :abilities
end
```

``` ruby
class Animal::Cat < Animal
  has_many :toys, foreign_key: :animal_id
end
```

and run `Animal::Cat.first.abilities` I get

```
ActiveRecord::AssociationNotFoundError: Association named 'abilities' was not found on Animal::Cat; perhaps you misspelled it?
```

which is super confusing. The other association works fine `Animal::Cat.first.toys`.
If sub-class contain no `has_many` associations then associations from super-class works fine.
If I remove `SUB_TYPES = [Animal::Cat].freeze` line – everything works as expected.

I believe this example should be considered as unrecognised circular dependency, but if might be also a bug with `has_many` and inheritance.

Please let me know if I any other info is needed. Thanks!

**Comment by mrageh** [comment]
HI @shlajin 

Can you please use https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_gem.rb to create a bug report?

**Comment by shlajin** [comment]
@mrageh bad news: I can not reproduce this bug with this template. Reason – Rails somehow allows me to have class `Animal` and classes like `Animal::Cat` in a same time. I can't rebuild this structure in plain Ruby (to refer `Animal::Cat` I have to define it before `Animal` class, but in this case I can't inherit from `Animal` as it should be first).

Also, I found that in case if I name my package Animal**s** (`Animals::Cat`), then circular dependency resolves correctly, so... this is not a big bug  :)

Anyway, I created a Rails repo with failing test – that's the most I can do at this point unfortunately. Run `rake test` to see test fails.
https://github.com/shlajin/Rails-4.2.5-AR-bug-demo

**Comment by al2o3cr** [comment]
@shlajin Referring to classes while they're still being constructed can cause weird things to happen - like this case. :)

I can get the test in your sample app to pass by removing the original `SUB_TYPES` line from `app/models/animal.rb` and adding this to the end of the file (*after* the end of the `Animal` class):

```ruby
Animal::SUB_TYPES = %w(Animal::Cat).map(&:constantize).freeze
```

**Comment by shlajin** [comment]
@al2o3cr nice workaround, thanks!

It does not blocks me anymore anyway (as I figured out why I get this strange things happen and how to avoid them), but this is certainly a bug :(
I'm not sure what to do from here: leave the issue open until someone who's capable of writing the fix close it eventually?

**Comment by matthewd** [comment]
Yeah this is definitely a bug (possibly two, depending on exactly what's happening).

@fxn it sounds like a circular dependency. Do you happen to know if it's a pattern that we just can't detect?

**Comment by fxn** [comment]
Let me understand the core of the issue with this minimal example:

```ruby
# app/models/animal.rb
class Animal
  Cat
end

# app/models/cat.rb
class Cat < Animal
end
```

Given an brand new app with only those two files:

```
$ bin/rails r Cat
... Circular dependency detected while autoloading constant Cat (RuntimeError)
```

Is that what's the ticket about?

**Comment by matthewd** [comment]
@fxn AIUI, that's the expected behaviour, but if you change `Cat` to `Animal::Cat`, you no longer get the error.

**Comment by fxn** [comment]
If we have

```ruby
# app/models/animal.rb
class Animal
  Cat
end

# app/models/animal/cat.rb
class Animal
  class Cat < Animal
  end
end
```

and you evaluate `Animal::Cat`, then first an autoloading for `Animal` is triggered. That one evaluates *animal.rb* and that file triggers autoloading of `Cat`, Rails locates *animal/cat.rb*, and at that point the `Animal` constant is already known, so we do not get the loop and autoloading `Animal` succeeds. Then, in the original call the `Cat` in `Animal::Cat` has already been defined as a side-effect and resolved by Ruby directly.

**Comment by fxn** [comment]
The check was introduced in https://github.com/rails/rails/commit/b33700f5580b4cd85379a1dc60fa341ac4d8deb2.

This particular case could in theory work, because in plain Ruby this warns but works:

```ruby
# animal.rb
class Animal
  require 'cat'
  Cat
end

# cat.rb
require 'animal'

class Cat < Animal
end
```

But I don't see a way to allow that autoload that does not evaluate *cat.rb* at least twice.

**Comment by fxn** [comment]
@shlajin one possible way to approach this would be the following.

```ruby
class Animal < ActiveRecord::Base
end

require_dependency 'cat'
```

and then instead of `SUB_TYPES` use `Animal.subclasses` or `Animal.descendants` depending on what you want.

The purpose of the `require_dependency` call is that subclasses/descendants are known regardless of execution path, but it needs to be maintained (just like `SUB_TYPES`).

**Comment by shlajin** [comment]
@fxn thanks! There are some workarounds to avoid this problem, but as for me I was super surprised when the Rails error was about misspelled `has_many` association and it took me a while to figure out what actually causes the problem (found a bug in relatively big project after refactoring). At this point I understand that I'm guilty because I used circular dependency and this ticket is mostly about appropriate error in this case. I made a mistake and I expect Rails to say "check your dependencies", not the "association was not found". I believe this could save a bunch of time of anybody who runs in a same problem; is that possible?

**Comment by fxn** [comment]
@shlajin I see.

The second example above has a `has_many :toys` that probably is not relevant. Could you please provide a minimal Rails application that reproduces the problem for us to play around?

**Comment by al2o3cr** [comment]
The part where associations in the parent aren't available in the descendant class also vaguely reminds me of a gotcha (mentioned in "Ruby Under the Microscope") where certain changes to included modules aren't reflected in places where the module was included. Need to find the details on that...

**Comment by fxn** [comment]
I can explain, because I actually suggested that section :).

Problem is Ruby linearizes and caches the ancestor chain of a class or module. If you add a method to an ancestor, the method is found because method lookup iterates the chain. But including a new module in an ancestor is not propagated to existing chains, so their methods are not found.

That behaviour is not coherent in my view, the ancestor chain is dynamic in Ruby by definition. Method lookup should (conceptually) recurse as needed with the ancestor chain as exists in that call.

Matz told me he would consider changing this if it did not impact performance, and if IIRC JRuby could very easily implement it. But it has not changed so far.

**Comment by shlajin** [comment]
@fxn sorry for the delay in reply.

Code in the repo (https://github.com/shlajin/Rails-4.2.5-AR-bug-demo) is the minimum example.
If subclass has no associations then inherited associations works fine. In other words, if I delete the line `has_many :toys` the bug will disappear (failing test will pass).

**Comment by fxn** [comment]
@shlajin awesome, thanks very much.

**Comment by fxn** [comment]
@shlajin hey, just a heads up to say I have not forgotten this issue but need to find a moment to sit down and investigate. Will come back to it.

**Comment by shlajin** [comment]
@fxn that's awesome, thanks! It does not block me though, but it's always nice to have less bugs in everyday framework :)

**Comment by justin808** [comment]
Aloha from Hawaii! I've got a possibly related issue with the latest Rails 4.2.x. I'm working on an app with heavy use of multi-level STI where we have parents, children, grandchildren, great-grandchildren.

The [Rails guides on auto loading with STI](http://guides.rubyonrails.org/autoloading_and_reloading_constants.html#autoloading-and-sti) advise to have the parent call `require_dependency` on all grandchildren and lower.

The problem we see is that if the first access is the to the child, then that autoloads the parent (via inheritance, and that call require_dependency on the grandchild, which autoloads the child, which autoloads the parent (circular dependency).

Copied from the guide.
```
A way to ensure this works correctly regardless of the order of execution is to load the leaves of the
tree by hand at the bottom of the file that defines the root class:

# app/models/polygon.rb
class Polygon < ActiveRecord::Base
end
require_dependency ‘square’

Only the leaves that are at least grandchildren need to be loaded this way. Direct subclasses do not
need to be preloaded. If the hierarchy is deeper, intermediate classes will be autoloaded recursively
from the bottom because their constant will appear in the class definitions as superclass.
```

By tracing in the autoloader, I could find the first controller that is autoloaded and then I can call require_dependency on a file that calls require_dependency on all top level STI parents. Besides including this in the controller, I had to do this require_dependency in "rails_helper" for rspec tests.

It seems that there should be some natural "hook" location to call this rather having to source dive and put in tracing to find the first controller loaded.

**Comment by fxn** [comment]
Hi Justin!

Note that the guide says specifically the "leaves" of the tree. The ones (leaves) which are grandchildren or lower. Leaves that are children of the root need no explicit loading.

**Comment by fxn** [comment]
@justin808 One possible solution would be to load the root constant in `ApplicationController`:

```ruby
class ApplicationController < AC::Base
  # Eager load the STI.
  Root
end
````

The problem there is that it would only work for requests, still the console and scripts run by `runner` could face this gotcha unless they also load `Root` upfront.

**Comment by justin808** [comment]
I tried calling `require_dependency` above ApplicationController and that does not work.

So is this a known issue @fxn?

Do you agree there should be a hook in the rails startup for the console, server, rake files, specs, etc. where I can specify a set of files to pre-load?

**Comment by fxn** [comment]
@justin808 it should work, in what sense it does not work?

**Comment by fxn** [comment]
@justin808 Regarding the hook, files in *config/initializers* have that role. The problem is that in development mode eager loading there is not enough because initializers are not loaded per request. So, when the server launches all is fine, then we edit something, reload, and the hierarchy that was loaded in the initializers is no longer reachable using the constants. They need to be reloaded againg.

I guess the guide could then advise to eager load in an initializer and also in any root class of the controller hierarchy.

Not elegant in my book :), problem is these features are kind of opposite to each other. While autoloading assumes a world in which things are loaded on-demand, STI assumes a world in which classes are known upfront.

**Comment by fxn** [comment]
@justin808 Let me summarize the steps, assuming `Root` is the root of the STI (typed off-the-top of my head.):

1. At the bottom of *root.rb* write a `require_dependency` call for each leave of the tree. Intermediate levels should not be included. (Leaves that are direct children can spare the call.)

2. Create *config/initializers/eager_load_root.rb* with a single line `require_dependency 'root'`.

3. At the top of `ApplicationController` and any other similar hiearchy root (rare, sometimes this happens for APIs or admin backends) add also `require_dependency 'root'`.

With that setup the line in *rails_helper.rb* should not be needed, referencing the intermediate child class in the first prompt of a console should work, `bin/rails runner Child` should work, and application development should work normally reloading also if things change.

**Comment by justin808** [comment]
@fxn I'll try that out right now. Is there anything that for sure should not inside of the initializers? and should I consider naming the initializer so this one loads either before or after the other initializers?

**Comment by justin808** [comment]
@fxn I re-arranged my initializers so that one for active_admin is at the very end (naming it `zz_active_admin.rb` and I'm naming my loader `z_eager_load_root.rb` so it comes right before that.

This seems to do the trick.

Any reason we still need:

> At the top of ApplicationController and any other similar hiearchy root (rare, sometimes this happens for APIs or admin backends) add also require_dependency 'root'.

I put some tracing in the code I notice that rails loads the file multiple times, first from the initializers and then from application_controller.

**Comment by justin808** [comment]
@fxn Maybe I see a more elegant solution that renaming the existing initializers to control load order:

http://guides.rubyonrails.org/configuring.html

> **config.after_initialize** takes a block which will be run after Rails has finished initializing the application. That includes the initialization of the framework itself, engines, and all the application's initializers in config/initializers. Note that this block will be run for rake tasks. Useful for configuring values set up by other initializers:

```ruby
config.after_initialize do
  ActionView::Base.sanitized_allowed_tags.delete 'div'
end
```

I look forward to your ideas. Thanks so much!

**Comment by fxn** [comment]
@justin808 Regarding the line in *application_controller.rb*, [`require_dependency` detects if a file has already been loaded](https://github.com/rails/rails/blob/0ae22e5d8db809f369335519251610f0459ca359/activesupport/lib/active_support/dependencies.rb#L367) even in development mode, so *root.rb* will be evaluated only once when the application boots and in development mode it won't be evaluated again as long as no file is edited.

The line is needed to still have this covered after a file is edited. When that happens, all autoloaded constants are deleted (including 'Root', 'Child', 'Grandchild'... the entire hierarchy), and that set of loaded files is cleared. When this happens, if the line was missing at this point the application would have no special handling of the STI. It would behave as if the initializer didn't exist, because the initializers do not run per request. That's the point, they only run at boot.

The `after_initialize` hook could be an alternative if you prefer it. I consider it a bit more framework-level though, at the application level *config/initializers* is the idiomatic place where apps put things to be run on initialization (and yes, if the order matters you do something to make it happen). Not that the application initializers are already run, so it would run *after* the active admin one.

**Comment by justin808** [comment]
@fxn Thanks!

With some tracing code, I notice the file of top level dependencies double loads in development mode for the application server, but that is probably harmless. I thought require_dependency would not load the same file, but definitely does in the case of requiring the same file from both an initializer, as well as ApplicationController.

Thanks again!

**Comment by fxn** [comment]
@justin808 that's strange. This is a minimal application that shows that the file is loaded as I described:

```
fxn@yeager:~/tmp/iload $ cat config/initializers/eager_load_root.rb 
require_dependency 'root'
fxn@yeager:~/tmp/iload $ cat config/routes.rb 
Rails.application.routes.draw do
  get '/', to: 'application#index'
end
fxn@yeager:~/tmp/iload $ cat app/controllers/application_controller.rb 
require_dependency 'root'

class ApplicationController < ActionController::Base
  def index
    head :ok
  end
end
fxn@yeager:~/tmp/iload $ cat app/models/root.rb 
p :LOADED
Root = 1
```

You'll see that launching the server prints ":LOADED" once. This is the one coming from the initializers. Then, issue a bunch of requests and you'll see the file is not loaded a second time. Then edit something and you'll see it loaded again.

If your application is schematically like that and it loads the file twice on boot we should investigate.

**Comment by justin808** [comment]
@fxn I'm creating a file called "autoloader_dependencies.rb". I'm missing a line like 

```
AutoloaderDependencies = 1
```

so that is defined. My file just contains a bunch of `require_dependency` calls.

**Comment by fxn** [comment]
In my example *app/models/root.rb* there is `Root = 1` because the example is minimal and defining a constant that matches the filename convention is enough. In a normal application that would be

```ruby
class Root < ActiveRecord::Base
  ...
end
```

if the model is persisted. I mean, the assignment is not important.

Where is the file *autoloader_dependencies.rb* located?

**Comment by justin808** [comment]
@fxn It's in app/models/autoloader_dependencies.rb.

**Comment by justin808** [comment]
@fxn I still get the file loaded twice (per trace statements), but now I get this warning as well.

```
App 7181 stderr: /app/models/autoloader_dependencies.rb:27: warning: already initialized constant AutoloaderDependencies
App 7181 stderr: /app/models/autoloader_dependencies.rb:27: warning: previous definition of AutoloaderDependencies was here
```

the file contains:
```
AutoloaderDependencies = 1
```

* Is `models` the right place to put this autoloader manifest?
* Is defining the constant to an int sufficient?

**Comment by justin808** [comment]
@fxn I found the issue. In my initializer, I used "require" and not "require_dependency". Thanks for the help!

**Comment by justin808** [comment]
Here's a tip for how I debugged this:

![2016-02-18_01-13-43](https://cloud.githubusercontent.com/assets/1118459/13141760/14be53bc-d5dd-11e5-9cbe-909e62b9e616.png)

**Comment by kellyjosephprice** [comment]
I would like to add that I had to use `config.to_prepare` instead of `config.after_initialize`. When eager loading, a child class could be loaded first.
